build-tools/packages/build-cli/src/handlers/checkFunctions.ts

```diff
@@ -923,29 +930,44 @@ export const checkCompatLayerGeneration: StateHandlerFunction = async (
 ): Promise<boolean> => {
 	if (testMode) return true;
 
-	const { context, bumpType } = data;
+	const { context, releaseGroup, bumpType } = data;
 
 	if (bumpType === "patch") {
 		log.verbose(`Skipping layer compat generation check for patch release.`);
 		BaseStateHandler.signalSuccess(machine, state);
 		return true;
 	}
 
-	// layerGeneration:gen should be run from the root. It will only update packages that have the layerGeneration:gen
-	// script defined in their package.json.
-	const result = await execa.command(`pnpm run -r layerGeneration:gen`, {
-		cwd: context.root,
-	});
-	log.verbose(result.stdout);
+	// Get packages to check based on release group or individual package
+	const packagesToCheck = isReleaseGroup(releaseGroup)
+		? context.packagesInReleaseGroup(releaseGroup)
+		: // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+			[context.fullPackageMap.get(releaseGroup)!];
 
-	// check for policy check violation
-	const gitRepo = await context.getGitRepository();
-	const afterPolicyCheckStatus = await gitRepo.gitClient.status();
-	const isClean = afterPolicyCheckStatus.isClean();
-	if (!isClean) {
+	// Check all packages using the validation logic
+	const packagesNeedingUpdate: { name: string; reason: string }[] = [];
+	
+	for (const pkg of packagesToCheck) {
+		// eslint-disable-next-line no-await-in-loop -- Need to check files sequentially
+		const result = await checkPackageCompatLayerGeneration(
+			pkg,
+			DEFAULT_GENERATION_DIR,
+			DEFAULT_GENERATION_FILE_NAME,
+			DEFAULT_MINIMUM_COMPAT_WINDOW_MONTHS,
+		);
+
+		if (result.needsUpdate) {
+			packagesNeedingUpdate.push({
+				name: pkg.name,
+				reason: result.reason,
+			});
+		}
+	}
```

Comment:
I think this pattern is common - take an iterable of packages and check them - maybe we ca change the signature of checkPackageCompatLayerGeneration to take an iterable of packages and when we need to pass a single package we just wrap it in an array?